File: src/org/genyris/io/parser/ParserXML.java
package org.genyris.io.parser;

import java.io.IOException;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.genyris.core.Constants;
import org.genyris.interp.Environment;
import org.genyris.core.Exp;
import org.genyris.core.Internable;
import org.genyris.core.PairSource;
import org.genyris.exception.GenyrisException;
import org.genyris.io.Parser;
import org.genyris.io.InStream;

public class ParserXML extends Parser {
    private SAXParser saxParser;
    private InStream inputStream;
    private Internable _symtab;
    private boolean _optionQname;

    public ParserXML(Internable table, InStream stream, boolean optionQname) throws GenyrisException {
        super(table, stream, Constants.DYNAMICSCOPECHAR2, Constants.CDRCHAR,
                Constants.COMMENTCHAR);
        inputStream = stream;
        _optionQname = optionQname;
        _symtab = table;
        try {
            SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
            saxParserFactory.setNamespaceAware(true);
            saxParser = saxParserFactory.newSAXParser();
        }
        catch (ParserConfigurationException | SAXException  e) {
            throw new GenyrisException(e.getMessage());
        }
    }


    @Override
    public Exp read(Environment env) throws GenyrisException {
        SAXHandler handler = new SAXHandler(env, _optionQname);
        try {
            saxParser.parse(new InputSource(inputStream.getReader()), handler);
        }
        catch (SAXException | IOException  e) {
            throw new GenyrisException(e.getMessage());
        }

        return handler.getTree();
    }

}

File: src/org/genyris/io/parser/ParserXML.kt
package org.genyris.io.parser

import org.genyris.core.Constants
import org.genyris.core.Exp
import org.genyris.core.Internable
import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import org.genyris.io.InStream
import org.genyris.io.Parser
import org.xml.sax.InputSource
import org.xml.sax.SAXException
import java.io.IOException
import javax.xml.parsers.ParserConfigurationException
import javax.xml.parsers.SAXParser
import javax.xml.parsers.SAXParserFactory

class ParserXML(private val _symtab: Internable?, private val inputStream: InStream) : Parser(
    _symtab, inputStream, Constants.DYNAMICSCOPECHAR2, Constants.CDRCHAR,
    Constants.COMMENTCHAR
) {
    private val saxParser: SAXParser? = null
    private val _optionQname = false

    init {
        try {
            val saxParserFactory = SAXParserFactory.newInstance()
            saxParserFactory.setNamespaceAware(true)
            saxParser = saxParserFactory.newSAXParser()
        } catch (e: ParserConfigurationException) {
            throw GenyrisException(e.getMessage())
        } catch (e: SAXException) {
            throw GenyrisException(e.getMessage())
        }
    }


    @kotlin.Throws(GenyrisException::class)
    override fun read(env: Environment?): Exp? {
        val handler = SAXHandler(env)
        try {
            val `is` = InputSource(inputStream.getReader())
            saxParser!!.parse(`is`, handler)
        } catch (e: SAXException) {
            throw GenyrisException(e.getMessage())
        } catch (e: IOException) {
            throw GenyrisException(e.getMessage())
        }

        return handler.getTree()
    }
}
